src/odemis/odemisd/main.py

```diff
@@ -321,78 +322,96 @@ def _terminate_all_alive(self):
         """
         mic = self._instantiator.microscope
-        alive = set(mic.alive.value)
-
-        # create a "graph" child->parents
-        parents = {c: set() for c in alive}  # comp -> set of comps (all its parents)
-        for comp in alive:
-            # TODO: if a component was created by delegation, use its creator
-            # instead of itself as "used" by the other components, to ensure
-            # they might not be terminated too early.
-            try:
-                for child in comp.children.value:
-                    # TODO: do not add the creator
-                    parents[child].add(comp)
-            except Exception:
-                logging.warning("Failed to find children of component %s when terminating",
-                                comp.name, exc_info=True)
-
-        # TODO: for the components created by delegation, either terminate them
-        # before their parents, or don't do it at all (as the parent normally
-        # is in charge of it).
 
-        # terminate all the components in order
-        while parents:
-            parent_less = tuple(c for c, p in parents.items() if not p)
-            if not parent_less:
-                # just pick a random component if some loop
-                logging.warning("All the components to terminate have parents: %s",
-                                parents)
-                parent_less = (tuple(parents.keys())[0],)
-
-            for c in parent_less:
-                cname = c.name
-                logging.debug("Stopping comp %s", cname)
-                # TODO: update the .alive VA every time a component is stopped?
-                # maybe not necessary as we are finishing _everything_
+        # alive component -> components which depends on it
+        self._dependents = {c: set() for c in set(mic.alive.value)}  # comp -> set of comps (all the comps that depend on it)
+        # Children will be terminated when their parents are terminated, so no need to keep them in .dependents
+        for comp in list(self._dependents.keys()):  # ._dependents will be modified, so we need a copy
+            for child in comp.children.value:
+                self._dependents.pop(child, None)
+        for comp in self._dependents:
+            deps = comp.dependencies.value
+            for dep in deps:
+                # if a component was created by delegation, use its creator
+                # instead of itself as "used" by the other components, to ensure
+                # they might not be terminated too early.
                 try:
-                    # FIXME: don't try to terminate the children created by
-                    # delegation after terminating their parent (and container)
-                    # For now it just causes a warning message, but it'd be
-                    # better to not have the warning at all.
-                    c.terminate()
+                    d = dep.parent.value if dep.parent else dep
```

Comment:
`.parent` is not a VA. It's a (special) attribute. So it should be accessed just as `dep.parent`. I think it should be possible to write simply `d = dep.parent or dep`. 
That said, I'm wondering whether what the current code does. Would it just always raise an exception for components with parents? If so, you'd see many of the warnings in the log. Or maybe it returned a `Pyro4.RemoteMethod` in this case (instead of a component).

Reply:
You're right, I didn't test this enough. `dep.parent.value` always returns a `Pyro4.RemoteMethod`. I'll change it. Do we still need the try/except block? The only thing that could happen that I can think of is that if `dep` is destroyed, accessing `dep.parent` will raise an AttributeError, but I couldn't reproduce it.

Reply:
never mind, your next comment already answers my question :)